File: problems/0601-0700/0662_cheapest-closing-hour/solution.ts
function cheapestClosingHour(customers: string): number {
    // penalty at closing hour j = (#'N' in customers[:j]) + (#'Y' in customers[j:])
    let prefixN = 0;
    let suffixY = 0;
    for (let i = 0; i < customers.length; i++) {
        if (customers[i] === "Y") suffixY++;
    }
    let bestJ = 0;
    let bestPenalty = prefixN + suffixY;
    for (let j = 1; j <= customers.length; j++) {
        if (customers[j - 1] === "N") prefixN++;
        else suffixY--;
        const penalty = prefixN + suffixY;
        if (penalty < bestPenalty) {
            bestPenalty = penalty;
            bestJ = j;
        }
    }
    return bestJ;
}

File: problems/0601-0700/0662_cheapest-closing-hour/solution.py
class Solution:
    def cheapestClosingHour(self, customers: str) -> int:
        # penalty at closing hour j = (#'N' in customers[:j]) + (#'Y' in customers[j:])
        prefix_n = 0
        suffix_y = sum(1 for c in customers if c == "Y")
        best_j = 0
        best_penalty = prefix_n + suffix_y
        for j in range(1, len(customers) + 1):
            if customers[j - 1] == "N":
                prefix_n += 1
            else:
                suffix_y -= 1
            penalty = prefix_n + suffix_y
            if penalty < best_penalty:
                best_penalty = penalty
                best_j = j
        return best_j
